nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,97 @@ with lib;
 let
 
+  # This type can accept three different types of expressions:
+  #   * A list of packages like `[ pkgs.foo pkgs.bar ]`
+  #   * An attribute set of packages like `{ inherit (pkgs) foo bar; }`
+  #   * A more structured attribute set of packages like `{ "foo".package = pkgs.foo.override { hasBar = false; }; }`
+  #
+  # The first two cases will be coerced into the third one, which allows the
+  # package to be disabled and its package and list of outputs to be overridden.
+  packagesType =
+    let
+      inherit (lib) types;
+
+      checkAttrsOfDrvs = attrs:
+        let
+          drvs = lib.attrValues attrs;
+        in
+        lib.isAttrs attrs &&
+        lib.length drvs > 0 &&
+        lib.all (drv: lib.isDerivation drv) drvs;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
+
+      mkOutputs =
+        let
+          mkEntry = output: lib.nameValuePair output true;
+        in
+        outputs: lib.listToAttrs (map mkEntry outputs);
+
+      drvToAttrs = drv:
+        let
+          defaultOutput = drv.${(lib.head (drv.outputs or [ "out" ]))}
+            or (throw "something is up with the outputs of this derivation: ${drv}");
+        in
+        {
+          ${defaultOutput.name} = {
+            package = defaultOutput;
+            outputs = mkOutputs (outputsFor drv);
+          };
+        };
+      drvToAttrsFolder = acc: drv: lib.recursiveUpdate acc (drvToAttrs drv);
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
+    in
+    types.coercedTo
+      (types.listOf types.package)
+      pkgsToAttrs
+      (types.coercedTo
+        # Compound nix types only do shallow type checking, but since here we
+        # want to distinguish between (attrsOf packagesByPrio) and (attrsOf submodule { ... })
+        # we need to actually look at the values to know in which case we are.
+        ((types.attrsOf types.package) // { check = checkAttrsOfDrvs; })
+        (lib.foldlAttrs (acc: _: drv: drvToAttrsFolder acc drv) {})
+        (types.attrsOf (types.submodule ( { config, lib, ... }: {
+          options = {
+            enable = (lib.mkEnableOption "the package") // {
+              default = true;
+            };
+            package = lib.mkOption {
+              type = types.package;
+              description = lib.mdDoc ''
+                The derivation that will be used for this package.
+              '';
+            };
+            outputs = lib.mkOption {
+              type = types.attrsOf types.bool;
+              description = lib.mdDoc ''
+                The outputs to install for this package.
+              '';
+            };
+          };
+
+          config = {
+            # We add the default outputs of the package by default.
+            # If a specific output has been selected, we only add that one.
+            outputs = mkOutputs (outputsFor config.package);
```

Comment:
Should this be a `mkDefault`?

Reply:
I don't think so because then we cannot add an output to an already included package, since it would overwrite the previously included outputs if they have lower priorities.

I wrote it like this so that we add on to the default outputs by default, if someone really want to override, they can do so with `mkForce`.